internal/controller/scheduler/scheduler.go

```diff
@@ -630,15 +630,31 @@ func (w *worker) Build(podSpec *corev1.PodSpec, skipCheckout bool, inputs buildI
 	// Set log collection grace period based on termination grace period.
 	// This allows the agent to coordinate log collection timing with pod termination.
 	if podSpec.TerminationGracePeriodSeconds != nil {
-		// Calculate log collection grace period: termination grace period minus buffer for cleanup
-		logCollectionGracePeriod := *podSpec.TerminationGracePeriodSeconds - 10
-		if logCollectionGracePeriod < 1 {
-			logCollectionGracePeriod = 1 // Minimum 1 second
-		}
-		agentContainer.Env = append(agentContainer.Env, corev1.EnvVar{
-			Name:  "BUILDKITE_KUBERNETES_LOG_COLLECTION_GRACE_PERIOD",
-			Value: fmt.Sprintf("%ds", logCollectionGracePeriod),
-		})
+		termGraceSecs := int(*podSpec.TerminationGracePeriodSeconds)
+		// When the agent cancels the job (e.g. when it receives SIGTERM), it
+		// first interrupts the job, waits for signal-grace-period, then
+		// terminates the job before waiting the remainder of the cancel-grace-
+		// period.
+		// When the pod is deleted, Kubernetes first sends SIGTERM to all
+		// containers. From Agent v3.110.0, kubernetes-bootstrap absorbs that
+		// signal and instead waits for the agent container to send an interrupt
+		// over the socket.
+		// Kubernetes will then take care of killing the job after
+		// TerminationGracePeriodSeconds is up. But we should still configure
+		// signal-grace-period so that the agent has time to upload logs and
+		// mark the job as finished.
+		signalGracePeriod := max(0, termGraceSecs-10)
+		cancelGracePeriod := max(termGraceSecs, signalGracePeriod+1)
```

Comment:
Since `signalGracePeriod = termGraceSecs - 10`, isn't this line effectively? This is kinda linked to my question below too.

```suggestion
		cancelGracePeriod := termGraceSecs
```

Reply:
It's an edge case, but has an effect. If `termGraceSecs` is 0 (valid, but seems silly to me to do that), then `signalGracePeriod` is 0 (not negative). But if `cancelGracePeriod` is also 0, then [the existing check in the agent](https://github.com/buildkite/agent/blob/a4dffdbed6b36623d52dfa1e2b3d677d6c53a96a/clicommand/cancel_signal.go#L63) will prevent it starting up.

Reply:
Done!